sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc

```diff
@@ -0,0 +1,1161 @@
+  {
+    // synchronize across the work-items that took both branches
+    sycl::group_barrier(inner_inner);
+  }
+}
+----
+
+
+=== Cluster-Groups
+
+A cluster-group is a contiguous collection of work-items created by subdividing
+a group into equally sized parts, such that each work-item is a member of
+exactly one cluster. The size of a cluster-group is a static (compile-time)
+property.
+
+
+==== Creation
+
+New cluster-groups are created by partitioning an existing group, using the
+`get_cluster_group` free-function.
+
+NOTE: Creating a cluster-group does not require synchronization of the parent
+group, since work-items can independently identify cluster members given a
+fixed cluster size.
+
+[source, c++]
+----
+namespace ext::oneapi::experimental {
+
+template <size_t ClusterSize, typename Group> 
+cluster_group<ClusterSize, Group> get_cluster_group(Group group);
+
+} // namespace ext::oneapi::experimental
+----
+
+_Constraints_: Available only if `sycl::is_group_v<std::decay_t<Group>> &&
+std::is_same_v<Group, sycl::sub_group>` is true. `ClusterSize` must be positive
+and a power of 2.
+
+_Preconditions_: `ClusterSize` must be less than or equal to the result of
+`group.get_max_local_range()`. `group.get_local_linear_range()` must be evenly
+divisible by `ClusterSize`.
+
+_Returns_: A `cluster_group<ClusterSize>` consisting of all work-items in
+`group` that are in the same cluster as the calling work-item.
+
+
+==== `cluster_group` Class
+
+[source, c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <std::size_t ClusterSize, typename ParentGroup>
+class cluster_group {
+public:
+  using id_type = id<1>;
+  using range_type = range<1>;
+  using linear_id_type = uint32_t;
+  static constexpr int dimensions = 1;
+  static constexpr sycl::memory_scope fence_scope = ParentGroup::fence_scope;
+    
+  id_type get_group_id() const;
+
+  id_type get_local_id() const;
+
+  range_type get_group_range() const;
+
+  range_type get_local_range() const;
+
+  linear_id_type get_group_linear_id() const;
+
+  linear_id_type get_local_linear_id() const;
+
+  linear_id_type get_group_linear_range() const;
+
+  linear_id_type get_local_linear_range() const;
+
+  bool leader() const;
+};
+
+}
+----
+
+NOTE: `cluster_group` is templated on a `ParentGroup` because it is expected
+that it will eventually be possible to construct a cluster-group from more
+than only sub-groups.
+
+[source,c++]
+----
+id_type get_group_id() const;
+----
+_Returns_: An `id` representing the index of the cluster-group.
+
+[source,c++]
+----
+id_type get_local_id() const;
+----
+_Returns_: An `id` representing the calling work-item's position within
+the cluster-group.
+
+[source,c++]
+----
+range_type get_group_range() const;
+----
+_Returns_: A `range` representing the number of cluster-groups.
+
+[source,c++]
+----
+range_type get_local_range() const;
+----
+_Returns_: A `range` representing the number of work-items in the cluster-group.
```

Comment:
Might be worth stating that this is always `ClusterSize`.  Should this function be `constexpr` too?

Reply:
`get_local_range()` isn't `constexpr` for any of the other group types.  Do you think it could be confusing to have some group types be `constexpr` and others not?

Reply:
I'm not sure.  I guess it's safer to leaver off the `constexpr`.  I think it would still be useful to note that this function always returns `ClusterSize`.